### PR TITLE
CMakeLists: warn if CMAKE_BUILD_TYPE is set/unset

### PR DESCRIPTION
**Pull request by batrick**
At some point the debug builds for wip branches no longer had the .git directory available so the Debug build type was unset. This meant we are no longer doing numerous checks (like mutex ownership checks) that we would normally be doing in the qa suite.

Related PR fixing the build process to always turn on these configs: https://github.com/ceph/ceph-build/pull/2167

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -24,9 +24,12 @@ endforeach()
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake/modules/")
 
 if(NOT CMAKE_BUILD_TYPE AND EXISTS "${CMAKE_SOURCE_DIR}/.git")
+  message(WARNING "Building with implicit CMAKE_BUILD_TYPE=Debug because .git exists.")
   set(default_build_type "Debug")
   set(CMAKE_BUILD_TYPE "${default_build_type}" CACHE
       STRING "Default BUILD_TYPE is Debug, other options are: RelWithDebInfo, Release, and MinSizeRel." FORCE)
+else()
+  message(WARNING "NOT building with implicit CMAKE_BUILD_TYPE=Debug because .git does NOT exist.")
```
Comment (idryomov) [inline CMakeLists.txt:32]: I get why you are adding this message but it seems a bit weird to me with two `NOT`s and also likely to go unnoticed. 

Given the history of repeated confusion and mismatched expectations in this area (226e614c95f1a1dcd79fa2011012ced1de19e6d3, etc), I would vote for making `CMAKE_BUILD_TYPE` required to be passed explicitly and fail the build (build files generation step) otherwise.

Comment (batrick) [reply CMakeLists.txt:32]: > I get why you are adding this message but it seems a bit weird to me with two `NOT`s and also likely to go unnoticed.

Can you suggest a different wording? I wanted the message to stand out (especially in jenkins logs).

> Given the history of repeated confusion and mismatched expectations in this area ([226e614](https://github.com/ceph/ceph/commit/226e614c95f1a1dcd79fa2011012ced1de19e6d3), etc), I would vote for making `CMAKE_BUILD_TYPE` required to be passed explicitly and fail the build (build files generation step) otherwise.

I like this idea but I expect many things may break with the automatic builds. I wonder what @tchaikov thinks (if you have time to comment Kefu :).

Comment (idryomov) [reply CMakeLists.txt:32]: > Can you suggest a different wording? I wanted the message to stand out (especially in jenkins logs).

`CMAKE_BUILD_TYPE not specified, assuming CMAKE_BUILD_TYPE=Debug because .git exists.`

`CMAKE_BUILD_TYPE not specified, assuming CMAKE_BUILD_TYPE=<what is assumed> because .git does NOT exist.`

Comment (batrick) [reply CMakeLists.txt:32]: Not sure from a cursory glance what the "default" even is because AFAICT the `CMAKE_BUILD_TYPE` is left unset from there on. There are some "Debug" checks in src/CMakeLists.txt but I don't see any for the other build types. I'm a bit confused.

Comment (idryomov) [reply CMakeLists.txt:32]: Since this PR has its roots in Jenkins, perhaps try to replicate locally what Jenkins does with `make-dist` such that `.git` isn't there, get to the build file generation step and run `grep CMAKE_BUILD_TYPE CMakeCache.txt` in the build directory?

Comment (tchaikov) [reply CMakeLists.txt:32]: > I get why you are adding this message but it seems a bit weird to me with two `NOT`s and also likely to go unnoticed.
> 

strictly speaking, it's not unusual to build a project without `CMAKE_BUILD_TYPE` from downstream maintainers' perspective. unlike regular users, they follow the guidelines and policies for packaging of their distro and packaging tools. for instance, both Debian and Fedora customize the building flags using CFLAGS, CXXFLAGS and LDFLAGS etc to be compliant to their specific policies. see https://www.debian.org/doc/manuals/debmake-doc/ch05.en.html#dh and https://docs.fedoraproject.org/en-US/packaging-guidelines/CMake/.

their packages built from CMake based project just **don't** use `CMAKE_BUILD_TYPE`, and they consider it a bug if the upstream sets a default `CMAKE_BUILD_TYPE`. as this practically forces the downstream maintainer to patch the `CMakeLists.txt` to drop whatever flags attached to the default `CMAKE_BUILD_TYPE`, as these flags are not always match the ones set by the distro.

in short, the upstream can be opinioned but its users should have the flexibility to opt-out if they want.

> Given the history of repeated confusion and mismatched expectations in this area ([226e614](https://github.com/ceph/ceph/commit/226e614c95f1a1dcd79fa2011012ced1de19e6d3), etc), I would vote for making `CMAKE_BUILD_TYPE` required to be passed explicitly and fail the build (build files generation step) otherwise.

based on the reasoning above. i'd vote nay. personally, i'd even suggest avoid printing warning if `CMAKE_BUILD_TYPE` is not set. as it is not a indication of misconfiguration.

Comment (idryomov) [reply CMakeLists.txt:32]: > their packages built from CMake based project just **don't** use `CMAKE_BUILD_TYPE`, and they consider it a bug if the upstream sets a default `CMAKE_BUILD_TYPE`. as this practically forces the downstream maintainer to patch the `CMakeLists.txt` to drop whatever flags attached to the default `CMAKE_BUILD_TYPE`, as these flags are not always match the ones set by the distro.
> 
> in short, the upstream can be opinioned but its users should have the flexibility to opt-out if they want.

I didn't suggest setting a default `CMAKE_BUILD_TYPE`.  Quite the opposite: I want to avoid any implicitness here and require passing `CMAKE_BUILD_TYPE` explicitly.

> 
> > Given the history of repeated confusion and mismatched expectations in this area ([226e614](https://github.com/ceph/ceph/commit/226e614c95f1a1dcd79fa2011012ced1de19e6d3), etc), I would vote for making `CMAKE_BUILD_TYPE` required to be passed explicitly and fail the build (build files generation step) otherwise.
> 
> based on the reasoning above. i'd vote nay. personally, i'd even suggest avoid printing warning if `CMAKE_BUILD_TYPE` is not set. as it is not a indication of misconfiguration.

I assume those who want to tweak compilation flags can just pass `None`:

```
CMAKE_BUILD_TYPE: A variable which controls the type of build
when using a single-configuration generator like the Makefile
generator. It is case-insensitive.

CMake will create by default the following variables when using a
single-configuration generator:

None (CMAKE_C_FLAGS or CMAKE_CXX_FLAGS used)
```
https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/Useful-Variables#compilers-and-tools

Comment (tchaikov) [reply CMakeLists.txt:32]: > > their packages built from CMake based project just **don't** use `CMAKE_BUILD_TYPE`, and they consider it a bug if the upstream sets a default `CMAKE_BUILD_TYPE`. as this practically forces the downstream maintainer to patch the `CMakeLists.txt` to drop whatever flags attached to the default `CMAKE_BUILD_TYPE`, as these flags are not always match the ones set by the distro.
> > in short, the upstream can be opinioned but its users should have the flexibility to opt-out if they want.
> 
> I didn't suggest setting a default `CMAKE_BUILD_TYPE`. Quite the opposite: I want to avoid any implicitness here and require passing `CMAKE_BUILD_TYPE` explicitly.

if we are in favor of this explictness. please make this explicit by sending a mail to ceph-maintainers. as the downstream maintainers should be aware of this -- they will have to pass `None` to `cmake`.

> 
> > > Given the history of repeated confusion and mismatched expectations in this area ([226e614](https://github.com/ceph/ceph/commit/226e614c95f1a1dcd79fa2011012ced1de19e6d3), etc), I would vote for making `CMAKE_BUILD_TYPE` required to be passed explicitly and fail the build (build files generation step) otherwise.
> > 
> > 
> > based on the reasoning above. i'd vote nay. personally, i'd even suggest avoid printing warning if `CMAKE_BUILD_TYPE` is not set. as it is not a indication of misconfiguration.
> 
> I assume those who want to tweak compilation flags can just pass `None`:
> 
> ```
> CMAKE_BUILD_TYPE: A variable which controls the type of build
> when using a single-configuration generator like the Makefile
> generator. It is case-insensitive.
> 
> CMake will create by default the following variables when using a
> single-configuration generator:
> 
> None (CMAKE_C_FLAGS or CMAKE_CXX_FLAGS used)
> ```
> 
> https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/Useful-Variables#compilers-and-tools

thanks. TIL.

Comment (batrick) [reply CMakeLists.txt]: I'm not prepared to make `CMAKE_BUILD_TYPE` required in this PR. I have edited the messages. Suggest we move the required setting to a tracker ticket for discussion (along with a mail to the maintainer list).

Comment (idryomov) [reply CMakeLists.txt]: Fine with me.

**Review by idryomov — APPROVED** [review]

**Comment by batrick** [comment]
> something went wrong here:
> 
> ```
> $ cmake -DCMAKE_BUILD_TYPE=Debug ..
> CMake Warning at CMakeLists.txt:32 (message):                  
>   CMAKE_BUILD_TYPE not specified, leaving unset because .git does NOT exist.
> ```

Ah, I see the issue. I'll write a followup fix.

**Comment by batrick** [comment]
https://github.com/ceph/ceph/pull/54272
